clang-tools-extra/clang-tidy/modernize/UseDesignatedInitializersCheck.cpp

```diff
@@ -0,0 +1,160 @@
+
+AST_MATCHER(CXXRecordDecl, isPOD) { return Node.isPOD(); }
+
+AST_MATCHER(InitListExpr, isFullyDesignated) {
+  if (const InitListExpr *SyntacticForm =
+          Node.isSyntacticForm() ? &Node : Node.getSyntacticForm()) {
+    return getNumberOfDesignated(SyntacticForm) == SyntacticForm->getNumInits();
+  }
+  return true;
+}
+
+AST_MATCHER(InitListExpr, hasMoreThanOneElement) {
+  return Node.getNumInits() > 1;
+}
+
+} // namespace
+
+UseDesignatedInitializersCheck::UseDesignatedInitializersCheck(
+    StringRef Name, ClangTidyContext *Context)
+    : ClangTidyCheck(Name, Context), IgnoreSingleElementAggregates(Options.get(
+                                         IgnoreSingleElementAggregatesName,
+                                         IgnoreSingleElementAggregatesDefault)),
+      RestrictToPODTypes(
+          Options.get(RestrictToPODTypesName, RestrictToPODTypesDefault)),
+      IgnoreMacros(
+          Options.getLocalOrGlobal(IgnoreMacrosName, IgnoreMacrosDefault)) {}
+
+void UseDesignatedInitializersCheck::registerMatchers(MatchFinder *Finder) {
+  const auto HasBaseWithFields =
+      hasAnyBase(hasType(cxxRecordDecl(has(fieldDecl()))));
+  Finder->addMatcher(
+      initListExpr(
+          hasType(cxxRecordDecl(RestrictToPODTypes ? isPOD() : isAggregate(),
+                                unless(HasBaseWithFields))
+                      .bind("type")),
+          IgnoreSingleElementAggregates ? hasMoreThanOneElement() : anything(),
+          unless(isFullyDesignated()))
+          .bind("init"),
+      this);
+}
+
+void UseDesignatedInitializersCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *InitList = Result.Nodes.getNodeAs<InitListExpr>("init");
+  const auto *Type = Result.Nodes.getNodeAs<CXXRecordDecl>("type");
+  if (!Type || !InitList) {
+    return;
+  }
+  const auto *SyntacticInitList = InitList->getSyntacticForm();
+  if (!SyntacticInitList) {
+    return;
+  }
+  std::optional<llvm::DenseMap<clang::SourceLocation, std::string>>
+      Designators{};
+  const auto LazyDesignators = [SyntacticInitList, &Designators] {
+    return Designators ? Designators
+                       : Designators.emplace(
+                             utils::getUnwrittenDesignators(SyntacticInitList));
+  };
+  const unsigned NumberOfDesignated = getNumberOfDesignated(SyntacticInitList);
+  if (0 == NumberOfDesignated) {
+    if (IgnoreMacros && InitList->getBeginLoc().isMacroID()) {
+      return;
+    }
+    if (SyntacticInitList->getNumInits() - NumberOfDesignated >
+        LazyDesignators()->size()) {
+      return;
+    }
+    {
+      DiagnosticBuilder Diag =
+          diag(InitList->getLBraceLoc(),
+               "use designated initializer list to initialize %0");
+      Diag << Type->getDeclName();
+      Diag << InitList->getSourceRange();
+      for (const Stmt *InitExpr : *SyntacticInitList) {
+        const std::string Designator =
```

Comment:
```suggestion
        const std::string& Designator =
```
No need to copy string